### PR TITLE
Refactor AWQ algo to enhance memory, computation and support foldng=False

### PR DESCRIPTION
**Pull request by xin3he**
## Type of Change

Enhancement

## Description

support folding=False for AWQ which includes self-absorb layers, namely MulLinear.

## Expected Behavior & Potential Risk

UT pass

## How has this PR been tested?

local test

### CODE REVIEW

**Review by lkk12014402 — APPROVED** [review]

**Review by wenhuach21 — APPROVED** [review]

File: neural_compressor/adaptor/torch_utils/teq.py [hunk 1]
```diff
@@ -207,7 +207,7 @@ def _scale_layer_weight(self, layer, scale):  ##input channel
         :param scale: The scale to be multiplied
         :return:
         """
-        if layer.__class__.__name__ == "TEQMulLinear":
+        if layer.__class__.__name__ == "MulLinear":
```
Comment (lkk12014402) [inline neural_compressor/adaptor/torch_utils/teq.py:210]: MulLinear and TEQMulLinear behave inconsistently, teq need to adapt

**Review by lkk12014402 — APPROVED** [review]

**Review by PenghuiCheng — APPROVED** [review]

File: neural_compressor/adaptor/torch_utils/awq.py [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+
+@torch.no_grad()
+def _get_act_scale(input_val):
+    tmp = [x.abs().view(-1, x.shape[-1]) for x in input_val]
+    tmp = torch.cat(tmp, dim=0)
+    return tmp.mean(0)
+
+
+class ActAwareWeightQuant:
+    """Implementation of Activation-aware Weight quantization (AWQ) algo."""
```
Comment (wenhuach21) [inline neural_compressor/adaptor/torch_utils/awq.py:92]: better add reference paper

File: neural_compressor/adaptor/torch_utils/awq.py [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+        self.model = model
+
+    def quantize(self, auto_scale=True, mse_range=True, folding=False, return_int=False):
+        """Execute AWQ quantization.
+
+        Args:
+            auto_scale (bool, optional): whether search scale. Defaults to True.
+            mse_range (bool, optional): whether search clip range. Defaults to True.
+            folding (bool, optional): whether only allow update scale when it can be fold 
+                                      to upper layer. Defaults to False.
```
Comment (wenhuach21) [inline neural_compressor/adaptor/torch_utils/awq.py:120]: previous layer?

File: neural_compressor/adaptor/torch_utils/awq.py [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+            # Step 1: fetch all input values of each linear for scale calculation
+            # use the first linear for QKV tuple
+            block_name = self.block_prefix + '.' + str(i)
+            block = fetch_module(self.model, block_name)
+            module_hook_config = {
+                v[0].split(block_name + '.')[1]: ['input'] for v in module_list
+            }
+            def block_calibration(model):
+                for args, kwargs in zip(self.total_block_args, self.total_block_kwargs):
+                    model(*args, **kwargs)
```
Comment (wenhuach21) [inline neural_compressor/adaptor/torch_utils/awq.py:151]: awq does not need any grad, you can add no_grad for all the forward

File: neural_compressor/adaptor/torch_utils/awq.py [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+        )
+        # Step 2: get block list and block prefix, number
+        self.block_prefix, self.block_num = get_block_prefix(model)
+        self.block_list = fetch_module(model, self.block_prefix)
+        self.bits = bits
+        self.group_size = group_size
+        self.scheme = scheme
+        self.sym_full_range = sym_full_range
+        self.weight_config = weight_config
+        self.model = model
```
Comment (wenhuach21) [inline neural_compressor/adaptor/torch_utils/awq.py:111]: you should call model.eval if it is not in eval mode

File: neural_compressor/adaptor/torch_utils/awq.py (hunk 4 above)
Comment (xin3he) [reply neural_compressor/adaptor/torch_utils/awq.py:151]: good point, thanks

File: neural_compressor/adaptor/torch_utils/awq.py (hunk 5 above)
Comment (xin3he) [reply neural_compressor/adaptor/torch_utils/awq.py:111]: ok

File: neural_compressor/adaptor/torch_utils/awq.py [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+def _get_act_scale(input_val):
+    tmp = [x.abs().view(-1, x.shape[-1]) for x in input_val]
+    tmp = torch.cat(tmp, dim=0)
+    return tmp.mean(0)
+
+
+class ActAwareWeightQuant:
+    """Implementation of Activation-aware Weight quantization (AWQ) algo."""
+    def __init__(self, model, example_inputs=None, calib_func=None, dataloader=None, n_samples=128,
+                 bits=4,  group_size=32, scheme='asym', sym_full_range=False, weight_config={},):
```
[the remark this answers is not in the document]
Comment (xin3he) [reply neural_compressor/adaptor/torch_utils/awq.py:94]: I remember 32 is the default value in our design meeting. I can change it for all algo if need.

File: neural_compressor/adaptor/torch_utils/awq.py (hunk 2 above)
Comment (xin3he) [reply neural_compressor/adaptor/torch_utils/awq.py:92]: The reference is in document.
